kvserver: load descriptor from disk in StateLoader.Load

tbg opened the pull request:
This is clean-up related to [#72222]. The range descriptor handling in the
replica state is somewhat awkward - callers loading the state are expected to
pass it in. This seems backwards; the descriptor is part of the state and so
the caller should pass in enough information to load it, but no more.

[#72222]: https://github.com/cockroachdb/cockroach/issues/72222. 

Release note: None

tbg committed:
kvserver: load descriptor from disk in StateLoader.Load

Release note: None

sumeerbhola reviewed (commented):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72229)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/stateloader/stateloader.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72229#-MnBw1KY94Ud9R8oI0fF:-MnBw1KY94Ud9R8oI0fG:b-yw7168) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cff6424cf83454b2e8dd663d7c16f4679403320/pkg/kv/kvserver/stateloader/stateloader.go#L56)):*
> ```Go
> // Load a ReplicaState from disk. The exception is the Desc field, which is
> // updated transactionally, and is populated from the supplied RangeDescriptor
> // under the convention that that is the latest committed version.
> ```

Does this comment need updating?


<!-- Sent from Reviewable.io -->

tbg commented:
Yes, this is still a draft. I'll let you know when it's ready for review. Appreciate the enthusiasm though :-)

tbg committed:
[wip]

Preventing uninitialized replicas from having a ReplicaState. This is
just playing whack-a-mole but it's a good exercise to learn what'll
go into separating the raft and the state part of a replica.

Release note: None
